In [1]:
# necessary packages #

#using Pkg
#Pkg.add("Distances")
using Distributions
using Random
using Distances
using LinearAlgebra
using SparseArrays
using IterativeSolvers
using ProgressMeter
#using NearestNeighbors # Can't use KDtree in julia for posterior prediction

In [2]:
using Pkg
Pkg.add("CUDAdrv")

  Updating registry at `~/.julia/registries/General`
  Updating git-repo `https://github.com/JuliaRegistries/General.git`
Fetching: [========================================>]  99.9 %0.0 %======>                                  ]  14.8 %]  27.3 %>                       ]  41.5 %Fetching: [=======================>                 ]  57.5 % [===============================>         ]  75.6 %88.0 % Resolving package versions...


Pkg.Types.ResolverError: Unsatisfiable requirements detected for package SIUnits [b9d75638]:
 SIUnits [b9d75638] log:
 ├─possible versions are: [0.0.1-0.0.6, 0.1.0] or uninstalled
 ├─restricted by compatibility requirements with Images [916415d5] to versions: [0.0.1-0.0.6, 0.1.0]
 │ └─Images [916415d5] log:
 │   ├─possible versions are: [0.2.0-0.2.50, 0.3.0-0.3.3, 0.4.0-0.4.50, 0.5.0-0.5.14, 0.6.0-0.6.1, 0.7.0, 0.8.0, 0.9.0-0.9.1, 0.10.0-0.10.1, 0.11.0-0.11.5, 0.12.0, 0.13.0, 0.14.0, 0.15.0-0.15.1, 0.16.0-0.16.1, 0.17.0-0.17.3, 0.18.0] or uninstalled
 │   └─restricted to versions 0.17.2 by an explicit requirement, leaving only versions 0.17.2
 └─restricted by julia compatibility requirements to versions: uninstalled — no versions left

In [3]:
Pkg.add("CUDAnative")

 Resolving package versions...


Pkg.Types.ResolverError: Unsatisfiable requirements detected for package SIUnits [b9d75638]:
 SIUnits [b9d75638] log:
 ├─possible versions are: [0.0.1-0.0.6, 0.1.0] or uninstalled
 ├─restricted by compatibility requirements with Images [916415d5] to versions: [0.0.1-0.0.6, 0.1.0]
 │ └─Images [916415d5] log:
 │   ├─possible versions are: [0.2.0-0.2.50, 0.3.0-0.3.3, 0.4.0-0.4.50, 0.5.0-0.5.14, 0.6.0-0.6.1, 0.7.0, 0.8.0, 0.9.0-0.9.1, 0.10.0-0.10.1, 0.11.0-0.11.5, 0.12.0, 0.13.0, 0.14.0, 0.15.0-0.15.1, 0.16.0-0.16.1, 0.17.0-0.17.3, 0.18.0] or uninstalled
 │   └─restricted to versions 0.17.2 by an explicit requirement, leaving only versions 0.17.2
 └─restricted by julia compatibility requirements to versions: uninstalled — no versions left

In [4]:
Pkg.add("CuArrays")

 Resolving package versions...


Pkg.Types.ResolverError: Unsatisfiable requirements detected for package SIUnits [b9d75638]:
 SIUnits [b9d75638] log:
 ├─possible versions are: [0.0.1-0.0.6, 0.1.0] or uninstalled
 ├─restricted by julia compatibility requirements to versions: uninstalled
 └─restricted by compatibility requirements with Images [916415d5] to versions: [0.0.1-0.0.6, 0.1.0] — no versions left
   └─Images [916415d5] log:
     ├─possible versions are: [0.2.0-0.2.50, 0.3.0-0.3.3, 0.4.0-0.4.50, 0.5.0-0.5.14, 0.6.0-0.6.1, 0.7.0, 0.8.0, 0.9.0-0.9.1, 0.10.0-0.10.1, 0.11.0-0.11.5, 0.12.0, 0.13.0, 0.14.0, 0.15.0-0.15.1, 0.16.0-0.16.1, 0.17.0-0.17.3, 0.18.0] or uninstalled
     └─restricted to versions 0.17.2 by an explicit requirement, leaving only versions 0.17.2

In [11]:
Pkg.add("SIUnits")

 Resolving package versions...


Pkg.Types.ResolverError: Unsatisfiable requirements detected for package TexExtensions [9b435220]:
 TexExtensions [9b435220] log:
 ├─possible versions are: [0.0.1-0.0.3, 0.1.0] or uninstalled
 ├─restricted by compatibility requirements with SIUnits [b9d75638] to versions: [0.0.1-0.0.3, 0.1.0]
 │ └─SIUnits [b9d75638] log:
 │   ├─possible versions are: [0.0.1-0.0.6, 0.1.0] or uninstalled
 │   └─restricted to versions * by an explicit requirement, leaving only versions [0.0.1-0.0.6, 0.1.0]
 └─restricted by julia compatibility requirements to versions: uninstalled — no versions left

In [6]:
]up

  Updating registry at `~/.julia/registries/General`
  Updating git-repo `https://github.com/JuliaRegistries/General.git`
 Resolving package versions...
 Installed MCMCChains ─────────── v0.3.8
 Installed IJulia ─────────────── v1.18.1
 Installed Images ─────────────── v0.18.0
 Installed Plots ──────────────── v0.24.0
 Installed Widgets ────────────── v0.6.1
 Installed Distributions ──────── v0.18.0
 Installed ImageCore ──────────── v0.8.2
 Installed ImageShow ──────────── v0.2.0
 Installed ImageDistances ─────── v0.2.1
 Installed ImageMorphology ────── v0.2.1
 Installed BinaryProvider ─────── v0.5.4
 Installed Compat ─────────────── v2.1.0
 Installed FileIO ─────────────── v1.0.6
 Installed PlotUtils ──────────── v0.5.8
 Installed PrettyTables ───────── v0.4.2
 Installed ImageAxes ──────────── v0.6.0
 Installed ColorVectorSpace ───── v0.7.0
 Installed MacroTools ─────────── v0.5.0
 Installed OrderedCollections ─── v1.1.0
 Installed PooledArrays ───────── v0.5.1
 Installed Optim ──────

In [119]:
include("util.j")

colnorm (generic function with 1 method)

In [120]:
# unnecessary packages #

#using Pkg
#Pkg.add("UnicodePlots")
using UnicodePlots   # check the structure of the sparse matrix
using BenchmarkTools

using StatsPlots
using MCMCChains
using PrettyTables

In [148]:
# set parameters
n1 = 1000
n2 = 4000
n3 = 8000
n4 = 16000
n5 = 100000
ϕ = 8;
δ = 0.1;
m = 5;

Random.seed!(1234);
coords1 = rand(2, n1);
coords2 = rand(2, n2);
coords3 = rand(2, n3);
coords4 = rand(2, n4);
coords5 = rand(2, n5);
ordx1 = sortperm(coords1[1, :]);  
ordx2 = sortperm(coords2[1, :]); 
ordx3 = sortperm(coords3[1, :]); 
ordx4 = sortperm(coords4[1, :]); 
ordx5 = sortperm(coords5[1, :]); 
coords1 = coords1[:, ordx1];
coords2 = coords2[:, ordx2];
coords3 = coords3[:, ordx3];
coords4 = coords4[:, ordx4];
coords5 = coords5[:, ordx5];
                                                        
NN1 = BuildNN(coords1, m, 0)
NN2 = BuildNN(coords2, m, 0); 
NN3 = BuildNN(coords3, m, 0); 
NN4 = BuildNN(coords4, m, 0); 
NN5 = BuildNN(coords5, m, 0); 
nnIndx_col1 = vcat(NN1.nnIndx, 1:n1);   
nnIndx_col2 = vcat(NN2.nnIndx, 1:n2); 
nnIndx_col3 = vcat(NN3.nnIndx, 1:n3); 
nnIndx_col4 = vcat(NN4.nnIndx, 1:n4); 
nnIndx_col5 = vcat(NN5.nnIndx, 1:n5); 
nnIndx_row1 = zeros(Int64, 0);  
nnIndx_row2 = zeros(Int64, 0); 
nnIndx_row3 = zeros(Int64, 0); 
nnIndx_row4 = zeros(Int64, 0); 
nnIndx_row5 = zeros(Int64, 0); 
for i in 2:m
    nnIndx_row1 = vcat(nnIndx_row1, fill(i, i-1))
    nnIndx_row2 = vcat(nnIndx_row2, fill(i, i-1))
    nnIndx_row3 = vcat(nnIndx_row3, fill(i, i-1))
    nnIndx_row4 = vcat(nnIndx_row4, fill(i, i-1))
    nnIndx_row5 = vcat(nnIndx_row5, fill(i, i-1))
end
nnIndx_row1 = vcat(nnIndx_row1, repeat((m + 1):n1, inner = m), 1:n1);
nnIndx_row2 = vcat(nnIndx_row2, repeat((m + 1):n2, inner = m), 1:n2);
nnIndx_row3 = vcat(nnIndx_row3, repeat((m + 1):n3, inner = m), 1:n3);
nnIndx_row4 = vcat(nnIndx_row4, repeat((m + 1):n4, inner = m), 1:n4);
nnIndx_row5 = vcat(nnIndx_row5, repeat((m + 1):n5, inner = m), 1:n5);

nIndx1 = length(NN1.nnIndx);
nIndx2 = length(NN2.nnIndx);
nIndx3 = length(NN3.nnIndx);
nIndx4 = length(NN4.nnIndx);
nIndx5 = length(NN5.nnIndx);
A1 = Array{Float64}(undef, nIndx1); D1 = Array{Float64}(undef, n1);
A2 = Array{Float64}(undef, nIndx2); D2 = Array{Float64}(undef, n2);
A3 = Array{Float64}(undef, nIndx3); D3 = Array{Float64}(undef, n3);
A4 = Array{Float64}(undef, nIndx4); D4 = Array{Float64}(undef, n4);
A5 = Array{Float64}(undef, nIndx5); D5 = Array{Float64}(undef, n5);
getAD(coords1, NN1.nnIndx, NN1.nnDist, NN1.nnIndxLU, ϕ, 0.5, A1, D1);
getAD(coords2, NN2.nnIndx, NN2.nnDist, NN2.nnIndxLU, ϕ, 0.5, A2, D2);
getAD(coords3, NN3.nnIndx, NN3.nnDist, NN3.nnIndxLU, ϕ, 0.5, A3, D3);
getAD(coords4, NN4.nnIndx, NN4.nnDist, NN4.nnIndxLU, ϕ, 0.5, A4, D4);
getAD(coords5, NN5.nnIndx, NN5.nnDist, NN5.nnIndxLU, ϕ, 0.5, A5, D5);

VK1 = Diagonal(1 ./ sqrt.(D1)) * (sparse(nnIndx_row1, nnIndx_col1, vcat(-A1, ones(n1))));
VK2 = Diagonal(1 ./ sqrt.(D2)) * (sparse(nnIndx_row2, nnIndx_col2, vcat(-A2, ones(n2))));
VK3 = Diagonal(1 ./ sqrt.(D3)) * (sparse(nnIndx_row3, nnIndx_col3, vcat(-A3, ones(n3))));
VK4 = Diagonal(1 ./ sqrt.(D4)) * (sparse(nnIndx_row4, nnIndx_col4, vcat(-A4, ones(n4))));
VK5 = Diagonal(1 ./ sqrt.(D5)) * (sparse(nnIndx_row5, nnIndx_col5, vcat(-A5, ones(n5))));

Xstar1 = sparse((n1 + 1):(2 * n1), 1:n1, fill(δ, n1), (2 * n1), n1)
Xstar2 = sparse((n2 + 1):(2 * n2), 1:n2, fill(δ, n2), (2 * n2), n2)
Xstar3 = sparse((n3 + 1):(2 * n3), 1:n3, fill(δ, n3), (2 * n3), n3)
Xstar4 = sparse((n4 + 1):(2 * n4), 1:n4, fill(δ, n4), (2 * n4), n4)
Xstar5 = sparse((n5 + 1):(2 * n5), 1:n5, fill(δ, n5), (2 * n5), n5)

Xstar1[1:n1, 1:n1] = VK1;
Xstar2[1:n2, 1:n2] = VK2;
Xstar3[1:n3, 1:n3] = VK3;
Xstar4[1:n4, 1:n4] = VK4;
Xstar5[1:n5, 1:n5] = VK5;

b = rand(Normal(), 2 * n5);

In [149]:
# check LSMR convergence rate

tt1 = lsmr(Xstar1, b[1:(2 * n1)], log = true); println(tt1[2])
tt2 = lsmr(Xstar2, b[1:(2 * n2)], log = true); println(tt2[2])
tt3 = lsmr(Xstar3, b[1:(2 * n3)], log = true); println(tt3[2])
tt4 = lsmr(Xstar4, b[1:(2 * n4)], log = true); println(tt4[2])
tt5 = lsmr(Xstar5, b[1:(2 * n5)], log = true); println(tt5[2])


Converged after 169 iterations.
Converged after 375 iterations.
Converged after 394 iterations.
Converged after 529 iterations.
Converged after 817 iterations.


In [150]:
# check preconditioner 
CondD1 = colnorm(Xstar1);
CondD2 = colnorm(Xstar2);
CondD3 = colnorm(Xstar3);
CondD4 = colnorm(Xstar4);
CondD5 = colnorm(Xstar5);
tt1 = lsmr(Xstar1 * Diagonal(1 ./ CondD1), b[1:(2 * n1)], log = true); println(tt1[2])
tt2 = lsmr(Xstar2 * Diagonal(1 ./ CondD2), b[1:(2 * n2)], log = true); println(tt2[2])
tt3 = lsmr(Xstar3 * Diagonal(1 ./ CondD3), b[1:(2 * n3)], log = true); println(tt3[2])
tt4 = lsmr(Xstar4 * Diagonal(1 ./ CondD4), b[1:(2 * n4)], log = true); println(tt4[2])
tt5 = lsmr(Xstar5 * Diagonal(1 ./ CondD5), b[1:(2 * n5)], log = true); println(tt5[2])

Converged after 103 iterations.
Converged after 170 iterations.
Converged after 198 iterations.
Converged after 235 iterations.
Converged after 329 iterations.


# summary:

# No.1

n1 = 1000
n2 = 4000
n3 = 8000
n4 = 16000
n5 = 100000
ϕ = 8;
δ = 1.0;
m = 5;


Converged after 46 iterations.
Converged after 89 iterations.
Converged after 90 iterations.
Converged after 137 iterations.
Converged after 282 iterations.

With precondition

Converged after 39 iterations.
Converged after 68 iterations.
Converged after 66 iterations.
Converged after 116 iterations.
Converged after 175 iterations.

# No.2
n1 = 1000
n2 = 4000
n3 = 8000
n4 = 16000
n5 = 100000
ϕ = 3;
δ = 1.0;
m = 5;

Converged after 68 iterations.
Converged after 129 iterations.
Converged after 129 iterations.
Converged after 197 iterations.
Converged after 392 iterations.

With precond

Converged after 52 iterations.
Converged after 88 iterations.
Converged after 85 iterations.
Converged after 143 iterations.
Converged after 206 iterations.

Conclusion: lower decay cause higher iteration number

# No.3
n1 = 1000
n2 = 4000
n3 = 8000
n4 = 16000
n5 = 100000
ϕ = 8;
δ = 0.1;
m = 5;

Converged after 169 iterations.
Converged after 375 iterations.
Converged after 394 iterations.
Converged after 529 iterations.
Converged after 817 iterations.

with precond:

Converged after 103 iterations.
Converged after 170 iterations.
Converged after 198 iterations.
Converged after 235 iterations.
Converged after 329 iterations.

Conclusion: lower δ higher iterations, the normalized precondition works

